gallery_dl/extractor/generic.py

```diff
@@ -0,0 +1,207 @@
+# -*- coding: utf-8 -*-
+
+"""Extractor for images in a generic web page."""
+
+from .common import Extractor, Message
+from .. import text
+import re
+import os.path
+
+
+class GenericExtractor(Extractor):
+    """Extractor for images in a generic web page."""
+
+    category = "generic"
+    directory_fmt = ("generic_{pageurl}",)
+    filename_fmt = "{filename}.{extension}"
+    archive_fmt = filename_fmt
+
+    # Allow any url, optionally prefixed by "g:" or "generic:"
+    # in case we want to override other extractors.
+    # Based on: https://tools.ietf.org/html/rfc3986#appendix-B
+    pattern = r"""(?ix)
+            (?P<generic>g(?:eneric)?:)?     # optional "g(eneric):" prefix
+            (?P<scheme>https?://)?          # optional http or https scheme
+            (?P<domain>[^/?&#]+)            # required domain
+            (?P<path>/[^?&#]*)?             # optional path
+            (?:\?(?P<query>[^/?#]*))?       # optional query
+            (?:\#(?P<fragment>.*))?$        # optional fragment
+            """
```

Comment:
I don't see any problems, the aim of this regexp is to detect the various URL components; it's up to the user to provide a valid URL (BTW, "localhost" or any host name in a local private network is a valid url); if the host can't be resolved, the user will receive an exit error from urllib, instead of 'No suitable extractor found'.

Re the URL scheme, currently it's optional for all extractors (except directlink.py), so why should be required here?
Another question could be what should it default to, currently I've set it to `https`, maybe it should be `http`?

Reply:
c2b8675 made the domain part of the regex stricter [-\w\.] (as it's used in many other extractors), and this is also fixes the failures in `test_add` and `test_add_module` (otherwise the "fake:foobar" url used by the FakeExtractor would be considered a valid domain for the generic extractor).
            
I also made sure the generic extractor is disabled by default, unless `extractor.generic.enabled` is set, mimicking the ytdl extractor, so maybe leaving the https scheme optional is not too much a problem, i.e. for a scheme-less url to work, one must enable the generic extractor, either via config file or prepending `g(eneric):` to the url.